Sources/SecureXPC/Codable Property Wrappers/File Descriptor XPC Wrappers.swift

```diff
@@ -0,0 +1,184 @@
+            throw DecodingError.dataCorrupted(context)
+        }
+        let closeOnEncode = try container.decode(Bool.self, forKey: CodingKeys.closeOnEncode)
+        
+        self = Self.init(descriptor: descriptor, closeOnEncode: closeOnEncode)
+    }
+}
+
+// MARK: Darwin file descriptor
+
+/// Wraps a Darwin file descriptor such that it can be sent over an XPC connection.
+///
+/// By default the provided file descriptor will be closed once it has been encoded.
+///
+/// When creating an ``XPCRoute`` that directly transfers this type as either the message or reply type, `DarwinFileDescriptorForXPC` must be the
+/// specified type, not `CInt`. This is not applicable when transferring a type which _contains_ a wrapped file descriptor as one of its properties.
+///
+/// This property wrapper shares an underlying representation with ``FileHandleForXPC`` and ``FileDescriptorForXPC`` and therefore may be used
+/// interchangeably between the server and client. However, to make use of such functionality requires routes with identical names and differing message and/or
+/// reply types.
+@propertyWrapper public struct DarwinFileDescriptorForXPC {
+    public var wrappedValue: CInt
+    fileprivate let closeOnEncode: Bool
+    
+    public init(wrappedValue: CInt, closeOnEncode: Bool = true) {
+        self.wrappedValue = wrappedValue
+        self.closeOnEncode = closeOnEncode
+    }
+}
+
+extension DarwinFileDescriptorForXPC: FileDescriptorCodable {
+    init(descriptor: CInt, closeOnEncode: Bool) {
+        self.wrappedValue = descriptor
+        self.closeOnEncode = closeOnEncode
+    }
+    
+    var descriptor: CInt {
+        wrappedValue
+    }
+    
+    func close() throws {
+        let closeResult = Darwin.close(wrappedValue)
+        guard closeResult == 0 else {
+            throw XPCError.unknown(description: """
+            Unable to close file descriptor: \(wrappedValue)
+            Result code: \(closeResult)
+            """)
+        }
+    }
+}
+
+// MARK: FileDescriptor
+
+/// Wraps a [`FileDescriptor`](https://developer.apple.com/documentation/system/filedescriptor) such that it can be sent over an XPC
+/// connection.
+///
+/// By default the provided file descriptor will be closed once it has been encoded.
+///
+/// When creating an ``XPCRoute`` that directly transfers this type as either the message or reply type, `FileDescriptorForXPC` must be the specified type,
+/// not `FileDescriptor`. This is not applicable when transferring a type which _contains_ a wrapped file descriptor as one of its properties.
+///
+/// This property wrapper shares an underlying representation with ``DarwinFileDescriptorForXPC`` and ``FileHandleForXPC`` and therefore may be
+/// used interchangeably between the server and client. However, to make use of such functionality requires routes with identical names and differing message and/or
+/// reply types.
+@available(macOS 11.0, *)
+@propertyWrapper public struct FileDescriptorForXPC {
```

Comment:
Would it be more explicit to wrap this with a `#if canImport(System)`? Not sure the pros/cons.

Reply:
It was introduced in macOS 11 as far as I can tell from the documentation, is this not the normal way to conditionally handle this on macOS? I've never done `canImport` for a package/framework that ships with an Apple OS (just optional 3rd party dependencies).

Reply:
> is this not the normal way to conditionally handle this on macOS?

I'm not sure!

Reply:
The one I have there was autosuggested by Xcode so I'll stick with it.